Hilt를 사용하여 의존성 주입하기

ParkChan opened the pull request:
### 프로그래밍 요구사항
- Hilt를 사용해서 의존성을 주입해주세요.
    - UseCase가 Inject되어있는 빈 ViewModel까지 만들어주세요.
        - ViewModel을 만들어야만 정상적으로 DI가 적용되서 빌드가 되는지 확인이 가능합니다.
- data 모듈에는 internal 키워드를 유지해주세요.
- app 모듈을 제외한 나머지 모듈에서는 hilt-core를 사용하시면 됩니다.
    - 참고 : https://mvnrepository.com/artifact/com.google.dagger/hilt-core

ParkChan committed:
[추가] DaggerHilt / Junit5 Gradle 추가

Dependencies.kt 수정

ParkChan committed:
[추가] Data Module DaggerHilt 적용

[@BeokBeok 가이드 참조]
app 모듈을 제외한 나머지 모듈에서는 hilt-core를 사용하시면 됩니다.
참고 : https://mvnrepository.com/artifact/com.google.dagger/hilt-core

ParkChan committed:
[추가] Timber 라이브러리 추가 및 Application Class 추가

ParkChan committed:
[추가] ViewModelScope, LifeCycleScope, LiveData 사용을 위한 디펜던시 추가

ParkChan committed:
[수정] Data Module DI로 변경하면서 Object -> class 로 변경

ParkChan committed:
[수정] Data Module : 통신 모듈 DI로 변경하면서 Object -> class 로 변경

ParkChan committed:
[추가] 코루틴 디펜던시 추가

ParkChan committed:
[추가] ViewModel LiveData 테스트를 위한 유틸 코드 추가

ParkChan committed:
[추가] ViewModel 추가, App Dagger Hilt 적용 및 뷰모델 테스트 코드 추가

빌드에러 발생

D:\android_workspace\github\parkchan\CleanArchitureStudy\app\build\generated\source\kapt\debug\com\chan\moviesearcher\MyApplication_HiltComponents.java:127: error: [Dagger/MissingBinding] com.chan.moviesearcher.domain.repository.MovieSearchRepository cannot be provided without an @Provides-annotated method.
  public abstract static class SingletonC implements MyApplication_GeneratedInjector,
                         ^
      com.chan.moviesearcher.domain.repository.MovieSearchRepository is injected at
          com.chan.moviesearcher.di.MovieSearchUseCaseModule.provideMovieSearchUseCase(repository)
      com.chan.moviesearcher.domain.MovieSearchUseCase is injected at
          com.chan.moviesearcher.viewmodel.MovieSearchViewModel(useCase)
      com.chan.moviesearcher.viewmodel.MovieSearchViewModel is injected at
          com.chan.moviesearcher.ScrollingActivity.viewModel
      com.chan.moviesearcher.ScrollingActivity is injected at
          com.chan.moviesearcher.ScrollingActivity_GeneratedInjector.injectScrollingActivity(com.chan.moviesearcher.ScrollingActivity) [com.chan.moviesearcher.MyApplication_HiltComponents.SingletonC → com.chan.moviesearcher.MyApplication_HiltComponents.ActivityRetainedC → com.chan.moviesearcher.MyApplication_HiltComponents.ActivityC]
  The following other entry points also depend on it:
      dagger.hilt.android.internal.lifecycle.HiltView

ParkChan committed:
[수정] 누락된 internal 추가

ParkChan committed:
[수정] dagger hilt 문제 수정중..

ParkChan committed:
[수정] Google Meet 원격 : dagger hilt 문제 수정 완료

가장 큰 부분으로
kapt DaggerHiltConfig.COMPILER 가 누락되어 있었고
fun 단위는 internal을 제거 하였으며

아래 클래스들의 inject 방식을
@Binds abstract 로 수정하였음.
MovieSearchModule
MovieSearchUseCaseModule

BeokBeok commented on app/build.gradle:
```diff
@@ -34,18 +40,40 @@ android {
     testImplementation TestConfig.ASSERTJ_CORE
     testImplementation TestConfig.MOCKK
 
+    androidTestImplementation TestConfig.ANDROID_CORE_TESTING
+    androidTestImplementation TestConfig.JUNIT_JUPITER_API
+    androidTestImplementation TestConfig.ASSERTJ_CORE
+    androidTestImplementation TestConfig.ANDROID_TEST_CORE
+    androidTestRuntimeOnly TestConfig.ANDROID_TEST_RUNNER
+    androidTestImplementation TestConfig.CORE
```
P2. 안드로이드 테스트(UI 테스트)는 JUnit5를 지원하지 않습니다

ParkChan replied:
혹시 아래 코드의 문제점을 말씀주신건가요?

```
package com.chan.moviesearcher

import androidx.test.platform.app.InstrumentationRegistry
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test


/**
 * Instrumented test, which will execute on an Android device.
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
class ExampleInstrumentedTest {
    @Test
    fun useAppContext() {
        // Context of the app under test.
        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
        assertEquals("com.chan.moviesearcher", appContext.packageName)
    }
}
```

ParkChan replied:
c70433e1c4d91cef78e9d341dc48e48b191104a2 해당클래스 삭제하였습니다.

안드로이드 테스트(UI 테스트)는 JUnit5를 지원하지 않는 부분은 몰랐던 부분인데
의견 감사드립니다.

BeokBeok commented on app/src/androidTest/java/com/chan/moviesearcher/ExampleInstrumentedTest.kt:
```diff
@@ -1,19 +1,15 @@
  *
  * See [testing documentation](http://d.android.com/tools/testing).
  */
-@RunWith(AndroidJUnit4::class)
```
P2. 미사용 테스트 코드는 제거해주세요

ParkChan replied:
c70433e1c4d91cef78e9d341dc48e48b191104a2 해당 클래스 삭제 하였습니다.

BeokBeok commented on app/src/main/java/com/chan/moviesearcher/viewmodel/MovieSearchViewModel.kt:
```diff
@@ -0,0 +1,32 @@
+    private val _throwable = MutableLiveData<Throwable>()
+    val throwable: LiveData<Throwable> get() = _throwable
+
+    fun getMovieList(start: Int, query: String) = viewModelScope.launch {
```
P2. 만약 해당 함수를 호출하는 도중에 화면을 회전하는 등의 라이프싸이클 관련하여 문제가 생기게 된다면, viewModelScope는 어떻게 처리가 될까요?

ParkChan replied:
492c2e60dd1e2eba65055f8f3d3951b6a4b19233 예외가 발생되어 앱이 정지될 수 있겠네요. 
try/catch 블럭으로 해결할 수도 있지만  CoroutineExceptionHandler 사용하여 
예외를 처리할 수 있도록 수정하였습니다.


[참조링크]https://developer.android.google.cn/kotlin/coroutines/coroutines-best-practices?hl=ko

예외에 주의
코루틴에서 발생하는 예외를 잘못 처리하면 앱이 비정상 종료될 수 있습니다. 예외가 발생할 수 있으면 viewModelScope 또는 lifecycleScope를 사용하여 만든 코루틴의 본문에서 예외를 포착합니다.

```
class LoginViewModel(
    private val loginRepository: LoginRepository
) : ViewModel() {

    fun login(username: String, token: String) {
        viewModelScope.launch {
            try {
                loginRepository.login(username, token)
                // Notify view user logged in successfully
            } catch (error: Throwable) {
                // Notify view login attempt failed
            }
        }
    }
}
```

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/util/InstantTaskExecutorExtension.kt:
```diff
@@ -0,0 +1,29 @@
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+
+class InstantTaskExecutorExtension : BeforeEachCallback, AfterEachCallback {
```
P1. src 폴더가 아니라 test 폴더로 이동시켜주세요

ParkChan replied:
app/src 경로가 아닌 app/test 경로를 말씀주신걸까요?

ParkChan replied:
1352307e9fd5daf1f8f977d19ad1628874105432 app의 build.gradle에 하위 항목을 추가하고 test폴더를 이동하였습니다😄

```
sourceSets {
        main.java.srcDirs += 'src/main/kotlin'
        test.java.srcDirs += 'test/kotlin'
        androidTest.java.srcDirs += 'src/androidTest/kotlin'
        androidTest.assets.srcDirs += files("$projectDir/schemas".toString())
    }
```

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/util/LiveDataTest.kt:
```diff
@@ -0,0 +1,48 @@
+ * Observes a [LiveData] until the `block` is done executing.
+ * [참조](https://medium.com/androiddevelopers/unit-testing-livedata-and-other-common-observability-problems-bb477262eb04)
+ */
+fun <T> LiveData<T>.observeForTesting(block: () -> Unit) {
```
P1. src 폴더가 아니라 test 폴더로 이동시켜주세요

ParkChan replied:
1352307e9fd5daf1f8f977d19ad1628874105432 이동시켰습니다!😄

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/util/MainCoroutineRule.kt:
```diff
@@ -0,0 +1,29 @@
+
+
+@ExperimentalCoroutinesApi
+class MainCoroutineRule(
```
P2. mockk를 사용하시면 CoroutineRule이 없이도 코루틴 테스트를 지원합니다

ParkChan replied:
좋네요!! 몰랐던 부분입니다.

ParkChan replied:
9f10c38efadc8111532e361f6167e80781bf0af7 MainCoroutineRule 파일 삭제하였습니다.

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/viewmodel/MovieSearchViewModelTest.kt:
```diff
@@ -0,0 +1,59 @@
+import org.junit.jupiter.api.extension.ExtendWith
+
+@ExtendWith(InstantTaskExecutorExtension::class)
+class MovieSearchViewModelTest {
```
P2. ExperimentalCoroutinesApi 어노테이션은 없어도 될 것 같아요

ParkChan replied:
a7290e664366c05f472c1946725b7b7a44681418 어노테이션 삭제하였습니다.

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/viewmodel/MovieSearchViewModelTest.kt:
```diff
@@ -0,0 +1,59 @@
+
+    @ExperimentalCoroutinesApi
+    @Test
+    @DisplayName("라이브 데이터 테스트 입니다")
```
P2. JUnit5에는 테스트 함수명이 한글로 가능하기에, DisplayName은 사용하지 않으셔도 됩니다.

ParkChan replied:
a7290e664366c05f472c1946725b7b7a44681418 의견 감사합니다. DisplayName 삭제하였습니다.

BeokBeok commented on app/src/test/kotlin/com/chan/moviesearcher/viewmodel/MovieSearchViewModelTest.kt:
```diff
@@ -0,0 +1,59 @@
+    @ExperimentalCoroutinesApi
+    @Test
+    @DisplayName("라이브 데이터 테스트 입니다")
+    fun `라이브 데이터 테스트`() = runBlocking {
```
P1. 테스트 함수 이름을 보다 직관적으로 작성해보면 어떨까요?
예를 들면 `영화 리스트를 불러옵니다` 와 같은 이름으로 작성할 수도 있을 것 같습니다.

ParkChan replied:
a7290e664366c05f472c1946725b7b7a44681418 
테스트 명이 직관적이지 못했네요! 직관적으로 변경해보았습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/di/MovieSearchModule.kt:
```diff
@@ -0,0 +1,29 @@
+
+    @Binds
+    @Singleton
+    abstract fun provideMovieDataSource(
```
P2. 함수명을 provides에서 binds로 변경해주세요

(나머지 binds를 사용하고 있는 파일들도 동일합니다)

ParkChan replied:
2443e62ab60510da1c6ddb24aedce7dbb8eff005 변경하였습니다~!

BeokBeok reviewed (changes requested):
작업하시느라 고생이 많으십니다.
약간의 코맨트를 작성하였습니다. :smile:

<details><summary>코드리뷰 우선순위 규칙 (P1~P3)</summary>
P1 > ★★★ : 적극적으로 고민해주세요 (Request changes)
<br>
코드 수정이 반드시 필요하다고 판단되는 경우입니다.<br>
리뷰 요청자께서 만약 수용할 수 없는 상황이라면 적합한 의견을 들어 토론해주세요.
<br></br>
P2 > ★★☆ : 웬만하면 반영해 주세요 (Comment)
<br>
코드 수정을 권장하거나 리뷰 요청자께서 해당 코드를 작성하신 의도를 알고자 하는 경우입니다.<br>
리뷰 요청자께서는 가급적이면 이유를 들어 설명해주시거나 의견을 달아주세요.
<br></br>
P3 > ★☆☆ : 반영해도 좋고 넘어가도 좋습니다 (Approve)
<br>
해당 의견을 반영하는 게 좋을지 고민해 보는 정도면 충분합니다.<br>
만약 리뷰 요청자께서 리뷰어의 의견을 반영하셨더라도 리뷰어에게 별도로 리뷰 요청을 드리지 않아도 됩니다.
</details>

ParkChan committed:
[PR/수정] P2. 안드로이드 테스트(UI 테스트)는 JUnit5를 지원하지 않습니다

ParkChan committed:
[PR/수정] P2. 만약 해당 함수를 호출하는 도중에 화면을 회전하는 등의 라이프싸이클 관련하여 문제가 생기게 된다면, viewModelScope는 어떻게 처리가 될까요?

ParkChan committed:
[PR/수정] P2. mockk를 사용하시면 CoroutineRule이 없이도 코루틴 테스트를 지원합니다

ParkChan committed:
[PR/수정] MovieSearchViewModelTest 수정
P2. 미사용 테스트 코드는 제거해주세요
P2. ExperimentalCoroutinesApi 어노테이션은 없어도 될 것 같아요
P1. 테스트 함수 이름을 보다 직관적으로 작성해보면 어떨까요?
예를 들면 영화 리스트를 불러옵니다 와 같은 이름으로 작성할 수도 있을 것 같습니다.
P2. JUnit5에는 테스트 함수명이 한글로 가능하기에, DisplayName은 사용하지 않으셔도 됩니다.

ParkChan committed:
[PR/수정] P2. 함수명을 provides에서 binds로 변경해주세요

기존 provide -> s 붙여줌

ParkChan committed:
[PR/수정] P1. src 폴더가 아니라 test 폴더로 이동시켜주세요

Unit Test, Android UI Test 폴더 sourceSets으로 지정

ParkChan committed:
[수정/테스트코드] java.lang.RuntimeException: Method myLooper in android.os.Looper not mocked.

build.gradle 에 아래와 같이 추가

android {
  testOptions {
    unitTests.returnDefaultValues = true
  }
}

BeokBeok commented on app/build.gradle:
```diff
@@ -12,12 +15,22 @@ android {
+        main.java.srcDirs += 'src/main/kotlin'
+        test.java.srcDirs += 'test/kotlin'
+        androidTest.java.srcDirs += 'src/androidTest/kotlin'
+        androidTest.assets.srcDirs += files("$projectDir/schemas".toString())
```
P3. 안드로이드 테스트는 필요 없을 것 같아요

ParkChan replied:
8e11af62514165ec2c60b5805268498f57e92cba UI 테스트는 사용하지 않으니 해당 코드 삭제하였습니다.

BeokBeok commented on app/src/main/kotlin/com/chan/moviesearcher/viewmodel/MovieSearchViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+    private val _throwable = MutableLiveData<Throwable>()
+    val throwable: LiveData<Throwable> get() = _throwable
+
+    private val handler = CoroutineExceptionHandler { _, exception ->
+        Timber.e(exception.message)
+    }
```
P3. 다른 개발자들이 봤을 때 handler 라는 명칭이 안드로이드의 handler와 혼동될 것 같습니다.
보다 명확한 이름으로 작성하시면 좋을 것 같습니다.

ParkChan replied:
e1b443fe1af61b620f5a10a630209b4822607f8a 혼동될 수 있겠네요. 보다 명확한 이름인 
coroutineExceptionHandler 로 수정 하였습니다.

BeokBeok commented on app/test/kotlin/com/chan/moviesearcher/util/LiveDataTest.kt:
```diff
@@ -0,0 +1,48 @@
+ * Observes a [LiveData] until the `block` is done executing.
+ * [참조](https://medium.com/androiddevelopers/unit-testing-livedata-and-other-common-observability-problems-bb477262eb04)
+ */
+fun <T> LiveData<T>.observeForTesting(block: () -> Unit) {
+    val observer = Observer<T> { }
+    try {
+        observeForever(observer)
+        block()
+    } finally {
+        removeObserver(observer)
+    }
+}
```
P3. 미사용 함수는 제거해주세요

ParkChan replied:
8b653b5f847e2404e33738fd28fdb2be3e018f31 미사용 함수 삭제하였습니다.

BeokBeok reviewed (approved):
<img src="https://user-images.githubusercontent.com/48344355/108810482-8fca5400-75ee-11eb-9a24-9931a3877f9d.png" width="30%">

마지막 단계까지 고생하셨습니다. 🎉 👏

이번 과정을 통해서 부디 많은 걸 얻어가셨길 바랍니다. 👍

ParkChan committed:
[PR/수정] P3. 안드로이드 테스트는 필요 없을 것 같아요

ParkChan committed:
[PR/수정] P3. 다른 개발자들이 봤을 때 handler 라는 명칭이 안드로이드의 handler와 혼동될 것 같습니다.
        보다 명확한 이름으로 작성하시면 좋을 것 같습니다.

ParkChan committed:
[PR/수정] P3. 미사용 함수는 제거해주세요
